File: tests/test_engine.nim
## The turn loop against a FAKE LLM transport.
##
## The central assertion is the batch shape: both seats' calls must go out
## together, so the fake records each call's in-flight window and the test
## asserts the two windows INTERSECT. A sequential implementation would still
## produce legal directives and pass every other test in this suite.

import std/[json, monotimes, os]
import lib/helpers


type
  Window = object
    seat: int
    startMs, endMs: int64

var windows: seq[Window]

proc nowMs(): int64 = getMonoTime().ticks div 1_000_000

proc parallelFake(reply: string, delayMs = 30): BatchFn =
  ## Answers every seat with the same body, holding each call open long enough
  ## for the windows to be meaningful.
  result = proc (calls: seq[BatchCall], timeoutSeconds: int): seq[BatchReply]
      {.closure, gcsafe.} =
    {.cast(gcsafe).}:
      let started = nowMs()
      sleep(delayMs)
      let ended = nowMs()
      for call in calls:
        windows.add Window(seat: call.seat, startMs: started, endMs: ended)
        result.add BatchReply(seat: call.seat, ok: true, text: reply)
    discard timeoutSeconds

proc llmSeats(engine: TurnEngine) =
  for seat in Seat:
    engine.policies[seat] = SeatPolicy(kind: pkLlm, prompt: "be brave",
      label: "test", connected: true)

const GoodReply = """{"note":"hold the shape","robots":[
 {"id":"AZ-1","role":"keeper","intent":"hold","target":[-17,0],"say":"arc"},
 {"id":"AZ-2","role":"striker","intent":"shoot","target":[3,1],"say":"go"},
 {"id":"AZ-3","role":"wing","intent":"intercept","target":[8,-4],"say":"wide"}]}"""

proc oneParallelBatch() =
  windows.setLen(0)
  var sim = playing(testConfig())
  var engine = newTurnEngine(nil, parallelFake(GoodReply))
  engine.llmSeats()
  engine.turn(sim, 0, 0)
  doAssert windows.len == 2,
    "expected one call per seat, saw " & $windows.len
  doAssert windows[0].seat != windows[1].seat, "the same seat was called twice"
  doAssert windows[0].startMs < windows[1].endMs and
    windows[1].startMs < windows[0].endMs,
    "the two seats' calls did not overlap: they were issued SEQUENTIALLY"
  for seat in Seat:
    doAssert sim.hasDirective[seat]
    doAssert sim.activeDirective[seat].source == dsLlm
    doAssert sim.stats[seat].llmTurns == 1
  var directives = 0
  for record in engine.records:
    if parseJson(record){"k"}.getStr() == "directive":
      inc directives
  doAssert directives == 2, "expected two directive records"
  report "both seats go out in ONE parallel batch and both directives install"

proc perTurnBudgetIsEnforced() =
  ## A hung client must not hold the turn past `turnBudgetMs`.
  var sim = playing(testConfig())
  sim.config.turnBudgetMs = 300
  sim.config.attempt1Ms = 200
  sim.config.retryMs = 100
  var engine = newTurnEngine(nil,
    proc (calls: seq[BatchCall], timeoutSeconds: int): seq[BatchReply]
        {.closure, gcsafe.} =
      {.cast(gcsafe).}:
        sleep(220)
      for call in calls:
        result.add BatchReply(seat: call.seat, error: "Timeout was reached"))
  engine.llmSeats()
  let started = nowMs()
  engine.turn(sim, 3, 0)
  let elapsed = nowMs() - started
  doAssert elapsed < 3000,
    "the turn ran " & $elapsed & " ms against a 300 ms budget"
  for seat in Seat:
    doAssert sim.activeDirective[seat].source == dsFallback,
      "a hung client did not fall back"
    doAssert sim.stats[seat].fallbackTurns >= 1
  var timeouts = 0
  for record in engine.records:
    let node = parseJson(record)
    if node{"k"}.getStr() == "fallback" and node{"cause"}.getStr() == "timeout":
      inc timeouts
  doAssert timeouts >= 2, "no timeout fallback records were written"
  report "a hung client is bounded by the per-turn budget and falls back"

proc exactlyOneRetry() =
  ## A timeout on attempt 1 buys exactly ONE retry; a second failure is the
  ## scripted fallback.
  var sim = playing(testConfig())
  var attempts = 0
  var engine = newTurnEngine(nil,
    proc (calls: seq[BatchCall], timeoutSeconds: int): seq[BatchReply]
        {.closure, gcsafe.} =
      {.cast(gcsafe).}:
        inc attempts
      for call in calls:
        if attempts == 1:
          result.add BatchReply(seat: call.seat, error: "Timeout was reached")
        else:
          result.add BatchReply(seat: call.seat, ok: true, text: GoodReply))
  engine.llmSeats()
  engine.turn(sim, 1, 0)
  doAssert attempts == 2, "expected exactly two batches, saw " & $attempts
  for seat in Seat:
    doAssert sim.activeDirective[seat].source == dsLlm,
      "the retry did not land"

  var always = newTurnEngine(nil,
    proc (calls: seq[BatchCall], timeoutSeconds: int): seq[BatchReply]
        {.closure, gcsafe.} =
      for call in calls:
        result.add BatchReply(seat: call.seat, ok: true, text: "no json here"))
  always.llmSeats()
  var sim2 = playing(testConfig())
  always.turn(sim2, 1, 0)
  for seat in Seat:
    doAssert sim2.activeDirective[seat].source == dsFallback,
      "two consecutive parse failures did not fall back"
  var causes = 0
  for record in always.records:
    let node = parseJson(record)
    if node{"k"}.getStr() == "fallback" and
        node{"cause"}.getStr() == "parse_error":
      inc causes
  doAssert causes == 4, "expected two attempts x two seats, saw " & $causes
  report "one timeout buys exactly one retry; two failures fall back"

proc budgetGuardFires() =
  ## The guard switches the LLM off for the rest of the match, so the episode
  ## ends complete/full_time rather than deadline.
  var sim = playing(testConfig())
  var calls = 0
  var engine = newTurnEngine(nil,
    proc (batch: seq[BatchCall], timeoutSeconds: int): seq[BatchReply]
        {.closure, gcsafe.} =
      {.cast(gcsafe).}:
        inc calls
      for call in batch:
        result.add BatchReply(seat: call.seat, ok: true, text: GoodReply))
  engine.llmSeats()
  # 680 s elapsed against a 690 s budget and a 9 s turn: 680 + 18 > 690.
  engine.turn(sim, 30, 680)
  doAssert engine.llmOff, "the budget guard did not fire"
  doAssert calls == 0, "the LLM was still called after the guard"
  var guards = 0
  for record in engine.records:
    if parseJson(record){"k"}.getStr() == "budget_guard":
      inc guards
  doAssert guards == 1, "no budget_guard record"
  engine.turn(sim, 31, 681)
  doAssert calls == 0, "the guard did not stick"
  for seat in Seat:
    doAssert sim.activeDirective[seat].source == dsFallback
  report "the budget guard fires once, sticks, and costs no network wait"

proc noCredentialsFallsBackInstantly() =
  ## With no client at all every turn falls back with NO network wait, which is
  ## what makes offline certification complete in seconds.
  var sim = playing(testConfig())
  var engine = newTurnEngine(nil, nil)
  engine.llmSeats()
  let started = nowMs()
  for turn in 0 ..< 40:
    engine.turn(sim, turn, 0)
  let elapsed = nowMs() - started
  doAssert elapsed < 2000,
    "40 credential-free turns took " & $elapsed & " ms"
  for seat in Seat:
    doAssert sim.stats[seat].fallbackTurns == 40
  report "40 credential-free turns settle in " & $elapsed & " ms"

proc scriptedSeatsNeverCallOut() =
  var sim = playing(testConfig())
  var calls = 0
  var engine = newTurnEngine(nil,
    proc (batch: seq[BatchCall], timeoutSeconds: int): seq[BatchReply]
        {.closure, gcsafe.} =
      {.cast(gcsafe).}:
        inc calls
      for call in batch:
        result.add BatchReply(seat: call.seat, ok: true, text: GoodReply))
  engine.policies[Azure] = SeatPolicy(kind: pkScripted, baseline: "formation")
  engine.policies[Crimson] = SeatPolicy(kind: pkLlm, prompt: "p")
  engine.turn(sim, 0, 0)
  doAssert calls == 1, "expected one batch, saw " & $calls
  doAssert sim.activeDirective[Azure].source == dsScripted
  doAssert sim.activeDirective[Crimson].source == dsLlm
  doAssert sim.stats[Azure].llmTurns == 0
  doAssert sim.stats[Crimson].llmTurns == 1
  report "a scripted seat never reaches the transport"

proc mercyAndWallClock() =
  ## Mercy fires at a goal difference of five; the wall-clock stop yields
  ## deadline/wall_clock; the physics guard yields fault/sim_fault with 0.5.
  var sim = playing(testConfig())
  sim.stats[Azure].goals = 5
  var ended = false
  for _ in 0 ..< 300:
    sim.stepIdle()
    if sim.phase == GameOver:
      ended = true
      break
  doAssert ended, "mercy never fired at a five-goal difference"
  doAssert sim.endRule == erMercy,
    "expected mercy, got " & endRuleText(sim.endRule)
  doAssert sim.endReason == reasonComplete

  var stopped = playing(testConfig())
  stopped.stepIdle(10)
  stopped.wallClockStop()
  doAssert stopped.endReason == reasonDeadline
  doAssert stopped.endRule == erWallClock
  doAssert stopped.phase == GameOver

  var faulted = playing(testConfig())
  faulted.finishGame(reasonFault, erSimFault)
  doAssert faulted.scorePermille(Azure) == 500
  doAssert faulted.scorePermille(Crimson) == 500
  doAssert not faulted.seatWon(Azure) and not faulted.seatWon(Crimson)
  report "mercy, the wall-clock stop and the physics fault all resolve"

proc seatsAlwaysScoreToOne() =
  for a in 0 .. 6:
    for b in 0 .. 6:
      var sim = playing(testConfig())
      sim.stats[Azure].goals = int32(a)
      sim.stats[Crimson].goals = int32(b)
      sim.finishGame(reasonComplete, erFullTime)
      doAssert sim.scorePermille(Azure) + sim.scorePermille(Crimson) == 1000,
        "scores do not sum to 1.0 at " & $a & "-" & $b
      doAssert sim.seatWon(Azure) == (a > b)
      doAssert sim.seatWon(Crimson) == (b > a)
  var pinned = playing(testConfig())
  pinned.stats[Azure].goals = 2
  pinned.stats[Crimson].goals = 1
  pinned.finishGame(reasonComplete, erFullTime)
  doAssert pinned.scorePermille(Azure) == 667, $pinned.scorePermille(Azure)
  doAssert pinned.scorePermille(Crimson) == 333
  report "the two seats' scores always sum to exactly 1.000"

proc disconnectKeepsPlaying() =
  ## A seat that drops mid-match keeps playing on the scripted layer, and
  ## revives on reconnect. No failure mode leaves a robot unactuated.
  var sim = playing(testConfig())
  var engine = newTurnEngine(nil, parallelFake(GoodReply))
  engine.llmSeats()
  engine.turn(sim, 0, 0)
  doAssert sim.activeDirective[Crimson].source == dsLlm
  sim.removePlayerAt(1)
  engine.policies[Crimson] = SeatPolicy(kind: pkScripted, baseline: "formation")
  engine.turn(sim, 1, 0)
  doAssert sim.activeDirective[Crimson].source == dsScripted
  let masks = sim.compileMasks(sim.activeDirective)
  var actuated = 0
  for i in 0 ..< RobotCount:
    if masks[i] != 0:
      inc actuated
  doAssert actuated > 0, "the dropped seat's trio went inert"
  engine.policies[Crimson] = SeatPolicy(kind: pkLlm, prompt: "back")
  engine.turn(sim, 2, 0)
  doAssert sim.activeDirective[Crimson].source == dsLlm,
    "the seat did not revive on reconnect"
  report "a dropped seat keeps playing and revives on reconnect"

when isMainModule:
  echo "test_engine"
  oneParallelBatch()
  exactlyOneRetry()
  perTurnBudgetIsEnforced()
  budgetGuardFires()
  noCredentialsFallsBackInstantly()
  scriptedSeatsNeverCallOut()
  mercyAndWallClock()
  seatsAlwaysScoreToOne()
  disconnectKeepsPlaying()
  echo "test_engine: all good"

File: tests/test_engine.py
"""The turn loop: batching, budgets, end conditions and degradation."""

from __future__ import annotations

import asyncio
import json
import time

import pytest
from conftest import StubSeat, make_config, require_sim_wasm

from cogball import defaults
from cogball.engine import TurnEngine, score_for
from cogball.llm import LlmClient
from cogball.replay import ReplayWriter
from cogball.sim import CogballSim

GOOD_REPLY = json.dumps({
    "note": "hold the shape",
    "robots": [{"id": rid, "role": "wing", "intent": "chase",
                "target": [0, 0], "pass_to": None, "kick": "auto",
                "say": "on it"} for rid in ("AZ-1", "AZ-2", "AZ-3")],
})


def reply_for(seat: int) -> str:
    ids = defaults.robot_ids_for_seat(seat)
    return json.dumps({
        "note": "hold the shape",
        "robots": [{"id": rid, "role": "wing", "intent": "chase",
                    "target": [0, 0], "pass_to": None, "kick": "auto",
                    "say": "on it"} for rid in ids],
    })


class RecordingClient(LlmClient):
    """Records the in-flight window of every call, per seat."""

    def __init__(self, delay=0.05, **kwargs):
        super().__init__(**kwargs)
        self.transport = "fake"
        self.disabled = False
        self._resolved = True
        self.delay = delay
        self.windows: list[tuple[float, float, str]] = []

    async def _complete(self, system, user):
        seat = 1 if "Magenta" in user else 0
        start = time.monotonic()
        await asyncio.sleep(self.delay)
        self.windows.append((start, time.monotonic(), user))
        return reply_for(seat)


class HungClient(LlmClient):
    def __init__(self, **kwargs):
        super().__init__(**kwargs)
        self.transport = "fake"
        self.disabled = False
        self._resolved = True

    async def _complete(self, system, user):
        await asyncio.sleep(3600)
        return ""


def disabled_client() -> LlmClient:
    client = LlmClient()
    client.disabled = True
    client._resolved = True
    return client


def build(config, seats, client, sim=None, monotonic=time.monotonic):
    require_sim_wasm()
    first = config.seed & 1
    sim = sim or CogballSim(seed=config.seed, first_kickoff_seat=first)
    writer = ReplayWriter(config, "unknown", first, {
        "players": [p.name for p in config.players],
        "aliases": list(defaults.ALIASES), "policy_kinds": [], "robots": []})
    engine = TurnEngine(sim, config, seats, client, writer,
                        monotonic=monotonic, first_kickoff_seat=first)
    return engine, writer


def llm_seats():
    return [StubSeat(0, "daveey", prompt="press high", scripted=None),
            StubSeat(1, "daveey-1", prompt="sit deep", scripted=None)]


async def test_both_seats_are_queried_in_one_parallel_batch():
    config = make_config(max_ticks=300)
    client = RecordingClient(delay=0.08)
    engine, _ = build(config, llm_seats(), client)
    await engine.run()

    assert len(client.windows) >= 4      # two turns x two seats
    a, b = client.windows[0], client.windows[1]
    overlap = min(a[1], b[1]) - max(a[0], b[0])
    assert overlap > 0, (
        "the two seats' calls did not overlap: they were issued "
        "sequentially, which is what blows the 720 s play budget")
    assert engine.llm_turns == [2, 2]


async def test_a_hung_coach_costs_at_most_the_turn_budget():
    config = make_config(max_ticks=150, turn_budget_seconds=0.4)
    engine, writer = build(config, llm_seats(), HungClient())
    started = time.monotonic()
    result = await engine.run()
    elapsed = time.monotonic() - started
    assert elapsed < 3.0, elapsed
    assert result.reason == "complete"
    assert result.fallback_turns == (1, 1)
    causes = [e for e in writer.events if e["type"] == "fallback"]
    assert causes and all(c["cause"] == "timeout" for c in causes)


async def test_the_budget_guard_finishes_the_match_scripted():
    """It must end complete/full_time, not deadline."""
    config = make_config(max_ticks=600, turn_budget_seconds=12,
                         wall_clock_budget_seconds=20)
    # A clock that is already 0.5 s from tripping `elapsed + 2*budget`.
    ticks = iter([0.0] + [1.0] * 100000)
    engine, writer = build(config, llm_seats(), HungClient(),
                           monotonic=lambda: next(ticks))
    result = await engine.run()
    assert result.reason == "complete"
    assert result.end_rule == "full_time"
    guards = [e for e in writer.events if e["type"] == "budget_guard"]
    assert len(guards) == 1 and guards[0]["turn"] == 0
    assert result.fallback_turns == (4, 4)
    assert all(c["budget_guard"] == 4 for c in result.fallback_causes)


async def test_the_wall_clock_hard_stop_reports_deadline():
    config = make_config(max_ticks=7200, wall_clock_budget_seconds=690)
    clock = iter([0.0, 0.0] + [10.0] * 20 + [700.0] * 100000)
    engine, _ = build(config, [StubSeat(0, "a"), StubSeat(1, "b")],
                      disabled_client(), monotonic=lambda: next(clock))
    result = await engine.run()
    assert result.reason == "deadline"
    assert result.end_rule == "wall_clock"
    assert result.final_tick < config.max_ticks


class FaultingSim:
    """Wraps the real sim and trips its fault flag after N ticks."""

    def __init__(self, inner, after: int, raise_instead=False):
        self.inner = inner
        self.after = after
        self.raise_instead = raise_instead
        self.ticks = 0

    def __getattr__(self, name):
        return getattr(self.inner, name)

    def step(self):
        self.ticks += 1
        if self.ticks > self.after and self.raise_instead:
            raise RuntimeError("wasmtime trap")
        self.inner.step()

    def fault(self):
        return self.ticks > self.after or self.inner.fault()


@pytest.mark.parametrize("raise_instead", [False, True])
async def test_a_sim_fault_scores_a_draw_and_still_writes_a_replay(
        raise_instead):
    config = make_config(max_ticks=900)
    require_sim_wasm()
    inner = CogballSim(seed=config.seed, first_kickoff_seat=0)
    sim = FaultingSim(inner, after=40, raise_instead=raise_instead)
    engine, writer = build(config, [StubSeat(0, "a"), StubSeat(1, "b")],
                           disabled_client(), sim=sim)
    result = await engine.run()
    assert result.reason == "fault"
    assert result.end_rule == "sim_fault"
    assert result.scores == (0.5, 0.5)
    assert result.winner is None
    assert 0 < writer.tick_count < config.max_ticks   # partial replay


class RunawaySim:
    """Wraps the real sim and reports a five-goal lead after N ticks."""

    def __init__(self, inner, after: int):
        self.inner = inner
        self.after = after
        self.ticks = 0

    def __getattr__(self, name):
        return getattr(self.inner, name)

    def step(self):
        self.ticks += 1
        self.inner.step()

    def goals(self, seat):
        if self.ticks > self.after:
            return defaults.MERCY_GOAL_DIFFERENCE if seat == 0 else 0
        return self.inner.goals(seat)


async def test_mercy_ends_a_runaway_match_at_a_turn_boundary():
    config = make_config(max_ticks=7200)
    require_sim_wasm()
    inner = CogballSim(seed=config.seed, first_kickoff_seat=0)
    engine, _ = build(config, [StubSeat(0, "a"), StubSeat(1, "b")],
                      disabled_client(), sim=RunawaySim(inner, after=200))
    result = await engine.run()
    assert result.reason == "complete"
    assert result.end_rule == "mercy"
    assert abs(result.goals[0] - result.goals[1]) >= \
        defaults.MERCY_GOAL_DIFFERENCE
    # the rules ended it, so it ended cleanly on a turn boundary
    assert result.final_tick % config.turn_ticks == 0
    assert result.final_tick < config.max_ticks


async def test_a_disconnected_seat_plays_formation_and_revives():
    config = make_config(max_ticks=900)
    seats = llm_seats()
    seats[1].connected = False
    client = RecordingClient(delay=0.0)
    engine, writer = build(config, seats, client)

    original = engine._collect_directives

    async def collect(turn, tick, world, views):
        if turn == 4:
            seats[1].connected = True     # the seat comes back
        await original(turn, tick, world, views)

    engine._collect_directives = collect
    result = await engine.run()

    assert result.llm_turns[0] == 6                # seat 0 never dropped
    # Three strikes of grace, then the seat degrades to the scripted layer;
    # it revives the turn after it reconnects.
    assert result.fallback_turns[1] == 2
    assert result.llm_turns[1] == 4
    sources = [e["source"] for e in writer.events
               if e["type"] == "directive" and e["seat"] == 1]
    assert sources == ["llm", "llm", "fallback", "fallback", "llm", "llm"]


async def test_scores_always_sum_to_exactly_one():
    for goals in [(0, 0), (1, 0), (3, 0), (5, 0), (2, 4), (7, 7)]:
        s0 = score_for(goals, 0)
        s1 = 1.0 - s0
        assert s0 + s1 == 1.0
        assert 0.0 <= s0 <= 1.0
    assert score_for((3, 0), 0) == 1.0
    assert score_for((1, 0), 0) == pytest.approx(0.6666666, abs=1e-6)
    assert score_for((0, 0), 0) == 0.5


async def test_every_turn_pushes_a_view_to_every_seat():
    config = make_config(max_ticks=600)
    seats = [StubSeat(0, "a"), StubSeat(1, "b")]
    engine, _ = build(config, seats, disabled_client())
    await engine.run()
    for seat in seats:
        assert len(seat.pushed) == 4
        view = seat.pushed[0]["view"]
        # anonymous aliases only: no real player name reaches a seat
        blob = json.dumps(seat.pushed)
        assert "daveey" not in blob
        assert view["you"]["alias"] in defaults.ALIASES
        assert set(view) == {
            "turn", "of", "clock", "score", "you", "pitch", "ball",
            "your_robots", "their_robots", "last_turn", "your_last_directive"}
        assert len(view["your_robots"]) == 3
        assert len(view["their_robots"]) == 3
        # the opponent's coaching is never visible
        assert "last_role" not in view["their_robots"][0]
